driver-core/src/main/java/com/datastax/driver/core/AbstractEventDebouncer.java

```diff
@@ -0,0 +1,212 @@
+import java.util.List;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * An abstract class to debounce events received by the Control Connection.
+ *
+ * This class accumulates received events and starts a new timer each time a new
+ * event is received, cancelling all pending timers.
+ *
+ * The events are only delivered:
+ * - if the current timer times out, in which case the events are delivered asynchronously;
+ * - if the queue is full, in which case the events are delivered synchronously.
+ */
+class AbstractEventDebouncer<T> {
+
+    private static final Logger logger = LoggerFactory.getLogger(AbstractEventDebouncer.class);
+
+    private final AtomicReference<DeliveryAttempt> pendingDelivery = new AtomicReference<DeliveryAttempt>(null);
+
+    private final ReentrantLock eventsLock = new ReentrantLock();
+
+    private final ReentrantLock startLock = new ReentrantLock();
+
+    private final Condition hasStarted = startLock.newCondition();
+
+    private final ScheduledExecutorService executor;
+
+    private final DeliveryCallback<T> callback;
+
+    private final long delayMs;
+
+    private final BlockingQueue<T> events;
+
+    private volatile boolean started = false;
+
+    AbstractEventDebouncer(ScheduledExecutorService executor, DeliveryCallback<T> callback, long delayMs, int maxPendingEvents) {
+        this.executor = executor;
+        this.callback = callback;
+        this.delayMs = delayMs;
+        this.events = new ArrayBlockingQueue<T>(maxPendingEvents);
+    }
+
+    void start() {
+        logger.trace("Starting debouncer...");
+        startLock.lock();
+        try {
+            started = true;
+            hasStarted.signalAll();
+        } finally {
+            startLock.unlock();
+        }
+        logger.trace("Debouncer started");
+        eventsLock.lock();
+        try {
+            if (!events.isEmpty()) {
+                logger.trace("Events were accumulated before the debouncer started: scheduling delivery now");
+                scheduleNewDelivery();
+            }
+        } finally {
+            eventsLock.unlock();
+        }
+    }
+
+    void stop() {
+        logger.trace("Stopping debouncer...");
+        started = false;
+        while (true) {
+            DeliveryAttempt previous = cancelPendingDelivery();
+            if (pendingDelivery.compareAndSet(previous, null)) {
+                break;
+            }
+        }
+        logger.trace("Debouncer stopped");
+    }
+
+    void eventReceived(T event) {
+        checkNotNull(event);
+        logger.trace("Event received {}", event);
+        enqueueEvent(event);
+        scheduleNewDelivery();
+    }
+
+    void deliverEvents() throws InterruptedException {
```

Comment:
The normal case is to have this method called by a thread in the scheduled executor. In case of back pressure, the following threads are also likely to execute this method:
* IO thread (nio worker) in Cluster.handle(Response), if the queue is full (is it a problem?)
* Client thread or cluster-reconnection pool in ControlConnection.tryConnect(Host, boolean) (only happens at initial connection attempt)
* cluster-worker (main executor) in maybeRefreshSchemaAndSignal()